File: go/libgo/nx/dns/dns.go
//
// weproxy@foxmail.com 2022/10/22
//

package dns

import (
	"errors"
	"fmt"
	"net"
	"strings"

	"golang.org/x/net/dns/dnsmessage"
)

// ...
var (
	ErrQueryNotFound = errors.New("not found")
	ErrQueryDropped  = errors.New("dropped")
	ErrQueryFaked    = errors.New("faked")
	ErrQueryHited    = errors.New("hited")
)

////////////////////////////////////////////////////////////////////////////////

var (
	_cache = newCacheMap()
)

////////////////////////////////////////////////////////////////////////////////

// Answer ...
type Answer struct {
	Msg *dnsmessage.Message
}

// Name ...
func (m *Answer) Name() string {
	if m.Msg != nil && len(m.Msg.Questions) > 0 {
		return strings.TrimSuffix(m.Msg.Questions[0].Name.String(), ".")
	}
	return ""
}

// Bytes ...
func (m *Answer) Type() dnsmessage.Type {
	if m.Msg != nil && len(m.Msg.Questions) > 0 {
		return m.Msg.Questions[0].Type
	}
	return dnsmessage.TypeA
}

// Bytes ...
func (m *Answer) Bytes() []byte {
	if m.Msg != nil {
		if b, err := m.Msg.Pack(); err == nil {
			return b
		}
	}
	return nil
}

////////////////////////////////////////////////////////////////////////////////

// OnRequest ...
func OnRequest(b []byte) (msg *dnsmessage.Message, answer *Answer, err error) {
	msg = &dnsmessage.Message{}
	if err = msg.Unpack(b); err == nil {
		answer, err = _cache.load(msg)
	}
	return
}

// OnResponse ...
func OnResponse(b []byte) (msg *dnsmessage.Message, err error) {
	msg = &dnsmessage.Message{}
	if err = msg.Unpack(b); err == nil {
		err = _cache.store(msg)
	}
	if err != nil {
		return nil, err
	}
	return
}

////////////////////////////////////////////////////////////////////////////////

// toAnswerString ...
func toAnswerString(arr []dnsmessage.Resource) string {
	var answs []string

	for _, answ := range arr {
		if answ.Body == nil {
			continue
		}
		switch answ.Header.Type {
		case dnsmessage.TypeA:
			answs = append(answs, fmt.Sprintf("%v", net.IP(answ.Body.(*dnsmessage.AResource).A[:])))
		case dnsmessage.TypeAAAA:
			answs = append(answs, fmt.Sprintf("%v", net.IP(answ.Body.(*dnsmessage.AAAAResource).AAAA[:])))
		case dnsmessage.TypeCNAME:
			answs = append(answs, strings.TrimRight(answ.Body.(*dnsmessage.CNAMEResource).CNAME.String(), "."))
		// case dnsmessage.TypeNS:
		// 	answs = append(answs, answ.Body.(*dnsmessage.NSResource).NS.String()+"NS")
		default:
			answs = append(answs, fmt.Sprintf("Type%v", int(answ.Header.Type)))
		}
	}

	return fmt.Sprintf("%v", answs)
}

File: cc/libcc/nx/dns/dns.cc
//
// weproxy@foxmail.com 2022/10/23
//

#include "dns.h"

#include "gx/fmt/fmt.h"
#include "gx/net/net.h"
#include "gx/strings/strings.h"
#include "logx/logx.h"

namespace nx {
namespace dns {

// ...
const error ErrQueryNotFound = errors::New("not found");
const error ErrQueryDropped = errors::New("dropped");
const error ErrQueryFaked = errors::New("faked");
const error ErrQueryHited = errors::New("hited");

namespace xx {
// ToString ...
string ToString(slice<Resource> arr) {
    stringz<> answs;

    for (int i = 0; i < len(arr); i++) {
        auto ans = arr[i];
        if (ans.Body == nil) {
            continue;
        }
        auto* p = ans.Body.get();
        switch (ans.Header.Type) {
            case Type::A:
                answs = append(answs, net::IP(((AResource*)p)->A).String());
                break;
            case Type::AAAA:
                answs = append(answs, net::IP(((AAAAResource*)p)->AAAA).String());
                break;
            case Type::CNAME:
                answs = append(answs, ((CNAMEResource*)p)->CNAME.String());
                break;
            // case Type::NS:
            //     answs = append(answs, ((NSResource*)p)->NS.String());
            default:
                answs = append(answs, fmt::Sprintf("Type%v", int(ans.Header.Type)));
                break;
        }
    }

    return GX_SS(answs);
}
}  // namespace xx

////////////////////////////////////////////////////////////////////////////////

// cacheLoad ...
extern R<Ref<Answer>, error> cacheLoad(Ref<Message> msg);

// cacheStore ...
extern error cacheStore(Ref<Message> msg);

////////////////////////////////////////////////////////////////////////////////

// Name ...
string Answer::Name() const {
    if (Msg && len(Msg->Questions) > 0) {
        return strings::TrimRight(Msg->Questions[0].Name.String(), ".");
    }
    return "";
}

// Type ...
Type Answer::Type() const {
    if (Msg && len(Msg->Questions) > 0) {
        return Msg->Questions[0].Type;
    }
    return Type::A;
}

// Bytes ...
bytez<> Answer::Bytes() const {
    if (Msg) {
        AUTO_R(b, err, Msg->Pack());
        if (err == nil) {
            return b;
        }
    }
    return {};
}

////////////////////////////////////////////////////////////////////////////////

// OnRequest ...
R<Ref<Message>, Ref<Answer>, error> OnRequest(bytez<> b) {
    auto msg = NewRef<Message>();
    auto err = msg->Unpack(b);
    if (err) {
        LOGS_E("[dns] err: " << err);
        return {nil, nil, err};
    }

    AUTO_R(answ, _, cacheLoad(msg));

    return {msg, answ, nil};
}

// OnResponse ...
R<Ref<Message>, error> OnResponse(bytez<> b) {
    auto msg = NewRef<Message>();
    auto err = msg->Unpack(b);
    if (err) {
        LOGS_E("[dns] err: " << err);
        return {nil, err};
    }

    err = cacheStore(msg);

    return {msg, nil};
}

}  // namespace dns
}  // namespace nx
